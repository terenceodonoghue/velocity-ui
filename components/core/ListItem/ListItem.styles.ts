import { css, FunctionInterpolation, Theme } from '@emotion/react';

const listItem: FunctionInterpolation<Theme> = ({ transitions }) => css`
  display: flex;
  align-items: center;
  justify-content: flex-start;
  height: 56px;
  padding: 0 0 0 25px;
  transition: background-color ${transitions.duration.standard}ms
    ${transitions.easing.sharp};
  cursor: pointer;
`;

export default listItem;
