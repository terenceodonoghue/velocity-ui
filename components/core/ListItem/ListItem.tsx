import React, { forwardRef } from 'react';
import { rgba } from 'polished';
import listItem from './ListItem.styles';
import type { ListItemProps } from './ListItem.types';

const ListItem = forwardRef<HTMLLIElement, ListItemProps>(
  ({ children, selected = false, ...props }, ref) => (
    <li
      css={[
        listItem,
        ({ palette }) => ({
          borderLeft: `solid 3px ${selected ? palette.accent : 'transparent'}`,
          backgroundColor: selected ? rgba(palette.accent, 0.1) : undefined,
          color: selected ? palette.accent : palette.neutral[500],
          '&:hover': {
            backgroundColor: !selected && rgba(palette.accent, 0.1),
          },
        }),
      ]}
      data-testid="listitem"
      ref={ref}
      {...props}
    >
      {children}
    </li>
  ),
);

export default ListItem;
