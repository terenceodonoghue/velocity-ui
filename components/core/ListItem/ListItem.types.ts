import { LiHTMLAttributes } from 'react';

export interface ListItemProps extends LiHTMLAttributes<HTMLLIElement> {
  selected?: boolean;
}
