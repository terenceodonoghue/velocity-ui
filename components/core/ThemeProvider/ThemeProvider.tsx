import React, { FunctionComponent } from 'react';
import { ThemeProvider as EmotionProvider, Theme } from '@emotion/react';
import merge from 'lodash/merge';
import { rgba } from 'polished';
import type { ThemeProps } from './ThemeProvider.types';

export const defaultTheme: Theme = {
  palette: {
    primary: rgba('#2e5bff', 0.2),
    secondary: '#e0e7ff',
    accent: '#2e5bff',
    neutral: {
      50: '#fff',
      100: '#f4f6fc',
      200: '#eef3f5',
      300: '#e4e8f0',
      400: '#bfc5d2',
      500: '#b0bac9',
      600: '#8798ad',
      700: '#8097b1',
      800: '#69707f',
      900: '#2E384D',
    },
    ui: {
      green: '#33ac2e',
      purple: '#8c54ff',
      red: '#d63649',
      teal: '#00c1d4',
      yellow: '#f7c137',
    },
    tonalOffset: 0.2,
  },
  transitions: {
    duration: {
      shortest: 150,
      shorter: 200,
      short: 250,
      standard: 300,
      complex: 375,
      enteringScreen: 225,
      leavingScreen: 195,
    },
    easing: {
      easeInOut: 'cubic-bezier(0.4, 0, 0.2, 1)',
      easeOut: 'cubic-bezier(0.0, 0, 0.2, 1)',
      easeIn: 'cubic-bezier(0.4, 0, 1, 1)',
      sharp: 'cubic-bezier(0.4, 0, 0.6, 1)',
    },
  },
  typography: {
    fontFamily: ['Rubik', 'sans-serif'].join(','),
    fontSize: 15,
    fontWeight: {
      light: 300,
      regular: 400,
      medium: 500,
      bold: 700,
    },
  },
  zIndex: {
    appBar: 1100,
    drawer: 1200,
  },
};

const ThemeProvider: FunctionComponent<ThemeProps> = ({ children, theme }) => (
  <EmotionProvider theme={theme ? merge(defaultTheme, theme) : defaultTheme}>
    {children}
  </EmotionProvider>
);

export default ThemeProvider;
