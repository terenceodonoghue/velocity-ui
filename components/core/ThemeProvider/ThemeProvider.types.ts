export interface Theme {
  palette?: {
    primary?: string;
    secondary?: string;
    accent?: string;
    neutral?: {
      50?: string;
      100?: string;
      200?: string;
      300?: string;
      400?: string;
      500?: string;
      600?: string;
      700?: string;
      800?: string;
      900?: string;
    };
    ui?: {
      green?: string;
      purple?: string;
      red?: string;
      teal?: string;
      yellow?: string;
    };
    tonalOffset?: number;
  };
  transitions?: {
    duration?: {
      shortest?: number;
      shorter?: number;
      short?: number;
      standard?: number;
      complex?: number;
      enteringScreen?: number;
      leavingScreen?: number;
    };
    easing?: {
      easeInOut?: string;
      easeOut?: string;
      easeIn?: string;
      sharp?: string;
    };
  };
  typography?: {
    fontFamily?: string;
    fontSize?: number;
    fontWeight?: {
      light?: number;
      regular?: number;
      medium?: number;
      bold?: number;
    };
  };
  zIndex?: {
    appBar?: number;
    drawer?: number;
  };
}

export interface ThemeProps {
  theme?: Theme;
}
