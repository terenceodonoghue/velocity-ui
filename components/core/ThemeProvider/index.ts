export { default, defaultTheme } from './ThemeProvider';
export type { Theme } from './ThemeProvider.types';
