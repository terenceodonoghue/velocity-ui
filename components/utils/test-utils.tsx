import React, { FunctionComponent, ReactElement } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import ThemeProvider from '~/components/core/ThemeProvider';

const Providers: FunctionComponent = ({ children }) => {
  return <ThemeProvider>{children}</ThemeProvider>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
): RenderResult => render(ui, { wrapper: Providers, ...options });

export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';

export { customRender as render };
