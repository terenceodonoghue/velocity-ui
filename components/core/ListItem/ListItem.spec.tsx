import React from 'react';
import { rgba } from 'polished';
import { render } from 'test-utils';
import ListItem from './ListItem';
import { defaultTheme } from '../ThemeProvider';

describe('ListItem', () => {
  it('is unselected by default', () => {
    const { container, getByTestId } = render(<ListItem />);

    expect(getByTestId('listitem')).toHaveStyle({
      borderLeft: 'solid 3px transparent',
    });
    expect(getByTestId('listitem')).not.toHaveStyle({
      backgroundColor: rgba(defaultTheme.palette.accent, 0.1),
    });
    expect(container).toMatchSnapshot();
  });

  it('renders selected item', () => {
    const { container, getByTestId } = render(<ListItem selected />);

    expect(getByTestId('listitem')).toHaveStyle({
      borderLeft: `solid 3px ${defaultTheme.palette.accent}`,
    });
    expect(getByTestId('listitem')).toHaveStyle({
      backgroundColor: rgba(defaultTheme.palette.accent, 0.1),
    });
    expect(container).toMatchSnapshot();
  });
});
