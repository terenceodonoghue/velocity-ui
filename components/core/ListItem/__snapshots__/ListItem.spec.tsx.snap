// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem is unselected by default 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  height: 56px;
  padding: 0 0 0 25px;
  -webkit-transition: background-color 300ms cubic-bezier(0.4, 0, 0.6, 1);
  transition: background-color 300ms cubic-bezier(0.4, 0, 0.6, 1);
  cursor: pointer;
  border-left: solid 3px transparent;
  color: #b0bac9;
}

.emotion-0:hover {
  background-color: rgba(46,91,255,0.1);
}

<div>
  <li
    class="emotion-0"
    data-testid="listitem"
  />
</div>
`;

exports[`ListItem renders selected item 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  height: 56px;
  padding: 0 0 0 25px;
  -webkit-transition: background-color 300ms cubic-bezier(0.4, 0, 0.6, 1);
  transition: background-color 300ms cubic-bezier(0.4, 0, 0.6, 1);
  cursor: pointer;
  border-left: solid 3px #2e5bff;
  background-color: rgba(46,91,255,0.1);
  color: #2e5bff;
}

<div>
  <li
    class="emotion-0"
    data-testid="listitem"
  />
</div>
`;
